sources/api/netdog/src/networkd_status.rs

```diff
@@ -0,0 +1,285 @@
+where
+    D: Deserializer<'de>,
+{
+    let address_vec: Vec<u8> = Deserialize::deserialize(deserializer)?;
+    ipaddr_from_vec(address_vec).map_err(D::Error::custom)
+}
+
+fn from_networkctl_addresses<'de, D>(deserializer: D) -> std::result::Result<Vec<IpAddr>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    #[derive(Debug, Deserialize)]
+    #[serde(rename_all = "PascalCase")]
+    struct NetworkctlAddress {
+        address: Vec<u8>,
+    }
+    let addresses: Vec<NetworkctlAddress> = Deserialize::deserialize(deserializer)?;
+    let mut addrs = Vec::new();
+    for addr in addresses.iter() {
+        addrs.push(ipaddr_from_vec(addr.address.clone()).map_err(D::Error::custom)?);
+    }
+    Ok(addrs)
+}
+
+impl NetworkDInterfaceStatus {
+    pub(crate) fn new(link: String) -> Result<Self> {
+        let systemd_networkctl_result = Command::new(NETWORKCTL)
+            .arg("status")
+            .arg("--json=pretty")
+            .arg(link)
+            .output()
+            .context(error::NetworkctlExecutionSnafu)?;
+        ensure!(
+            systemd_networkctl_result.status.success(),
+            error::FailedNetworkctlSnafu {
+                stderr: String::from_utf8_lossy(&systemd_networkctl_result.stderr)
+            }
+        );
+
+        let status_output = String::from_utf8(systemd_networkctl_result.stdout)
+            .context(error::NetworkctlParsingSnafu {})?;
+
+        let networkd_status =
+            serde_json::from_str(&status_output).context(error::NetworkctlDeserializeSnafu {})?;
+
+        Ok(networkd_status)
+    }
+
+    // Fetches the IP Address for the primary interface. If there are no addresses, this is
+    // an error. If there is one, it should be returned. If there is more than one, then find
+    // the first IPv4 address and return it, if there are none, return the first IPv6 address.
+    pub(crate) fn primary_address(&self) -> Result<IpAddr> {
+        // Find IPv4 first
+        for addr in self.addresses.iter() {
+            if addr.is_ipv4() {
+                return Ok(*addr);
+            }
+        }
+        // No IPv4 addresses, then attempt to return the first IPv6, otherwise return error
+        match self.addresses.first() {
+            Some(addr) => Ok(*addr),
+            None => error::NoIpAddressSnafu {
+                interface: self.name.clone(),
+            }
+            .fail(),
+        }
+    }
+}
+
+mod error {
+    use crate::interface_id::InterfaceId;
+    use snafu::Snafu;
+    use std::{io, string::FromUtf8Error};
+
+    #[derive(Debug, Snafu)]
+    #[snafu(visibility(pub(crate)))]
+    pub(crate) enum Error {
+        #[snafu(display("Failed to run 'networkctl': {}", source))]
+        NetworkctlExecution { source: io::Error },
+
+        #[snafu(display("'networkctl' failed: {}", stderr))]
+        FailedNetworkctl { stderr: String },
+
+        // Consider maybe shifting to length only for this error "Expected a Vec of length 4 but it was {}", v.len()
```

Comment:
Seems like this comment is leftover from development - I believe your error message covers this case.